In [43]:
!pip install biopandas
!pip install py3Dmol
!pip install ipywidgets

     |████████████████████████████████| 121 kB 26.9 MB/s eta 0:00:01
     |████████████████████████████████| 1.6 MB 24.9 MB/s eta 0:00:01
     |████████████████████████████████| 243 kB 68.0 MB/s eta 0:00:01
     |████████████████████████████████| 9.9 MB 65.8 MB/s eta 0:00:01
     |████████████████████████████████| 57 kB 9.3 MB/s  eta 0:00:01
     |████████████████████████████████| 86 kB 5.7 MB/s  eta 0:00:01


In [1]:
!rm qmu*

In [ ]:
import networkx as nx
from utility.MoleculeParser import MoleculeData
from utility.QMUQUBO import QMUQUBO
from utility.AnnealerOptimizer import Annealer
from utility.ResultProcess import ResultParser
import time

timestamp = time.strftime("%Y%m%d-%H")
%matplotlib inline

In [ ]:
# initial parameters for experiment data
s3_bucket = f"amazon-braket-1a222675c751" # the name of the bucket
prefix = "annealer-experiment" # the name of the folder in the bucket

raw_path = './molecule-data/Aspirin.mol2' # the mol2 file for this experiment

mol_data = MoleculeData(raw_path, 'qmu')

data_path = mol_data.save("latest")

num_rotation_bond = mol_data.bond_graph.rb_num
print(f"You have loaded the raw molecule data and saved as {data_path}. \n\
This molecule has {num_rotation_bond} rotable bond")

In [3]:
# initial the QMUQUBO object
init_param = {}
method = ['pre-calc']

for mt in method:
    if mt == 'pre-calc':
        init_param[mt] = {}
        init_param[mt]['param'] = ['M', 'D', 'A', 'hubo_qubo_val']
    
qmu_qubo = QMUQUBO(mol_data, method, **init_param)

INFO:root:initial pre-calculate for constructing molecule QUBO


In [4]:
# set the parameters for model
model_param = {}
# parameters
num_rotation_bond = mol_data.bond_graph.rb_num

method = 'pre-calc'
model_param[method] = {}
# model_param[method]['M'] = range(1, num_rotation_bond+1)
model_param[method]['M'] = [2,3]
model_param[method]['D'] = [2,4,8]
model_param[method]['A'] = [300]
model_param[method]['hubo_qubo_val'] = [200]

qmu_qubo.build_model(**model_param)

INFO:root:Construct model for M:2,D:2,A:300,hubo_qubo_val:200 0.00010190407435099284 min
INFO:root:Construct model for M:2,D:4,A:300,hubo_qubo_val:200 0.0003571073214213053 min
INFO:root:Construct model for M:2,D:8,A:300,hubo_qubo_val:200 0.0009239514668782552 min
INFO:root:Construct model for M:3,D:2,A:300,hubo_qubo_val:200 0.00020002126693725587 min
INFO:root:Construct model for M:3,D:4,A:300,hubo_qubo_val:200 0.0015513022740681966 min
INFO:root:Construct model for M:3,D:8,A:300,hubo_qubo_val:200 0.009593506654103598 min


0

In [6]:
qmu_qubo.model_info[method]

{'M': set(), 'D': set(), 'A': set(), 'hubo_qubo_val': set()}

In [7]:
# save the model
model_path = qmu_qubo.save("latest")

print(f"You have built the QUBO model and saved it as {model_path}")

INFO:root:finish save qmu_Aspirin_model_latest.pickle


You have built the QUBO model and saved it as ./qmu_Aspirin_model_latest.pickle


In [8]:
qmu_qubo_optimize = QMUQUBO.load(model_path)

In [9]:
# get the model you want to optimize
M = 3
D = 2
A = 300
hubo_qubo_val = 200
model_name = "{}_{}_{}_{}".format(M, D, A, hubo_qubo_val)
method = "pre-calc"

qubo_model = qmu_qubo_optimize.get_model(method, model_name)

In [19]:
qubo_model

{'qubo': defaultdict(float,
             {('x_3_1', 'x_3_2'): 600,
              ('x_3_1', 'x_2_1'): 200.0,
              ('x_3_1', 'x_3_1*x_2_1'): -400.0,
              ('x_3_1', 'x_2_2'): 200.0,
              ('x_3_1', 'x_2_2*x_3_1'): -400.0,
              ('x_3_2', 'x_2_1'): -4.2291767363180615,
              ('x_3_2', 'x_2_2'): -4.07974161165229,
              ('x_3_2', 'x_1_1'): 200.0,
              ('x_3_2', 'x_3_2*x_1_1'): -400.0,
              ('x_3_2', 'x_1_2'): 200.0,
              ('x_3_2', 'x_1_2*x_3_2'): -400.0,
              ('x_2_1', 'x_2_2'): 600,
              ('x_2_1', 'x_1_1'): -4.04507719278082,
              ('x_2_1', 'x_1_2'): -4.045032546085246,
              ('x_2_1', 'x_3_1*x_2_1'): -400.0,
              ('x_2_1', 'x_3_2*x_1_1'): -5.422469200549025,
              ('x_2_1', 'x_1_2*x_3_2'): -5.422444941277003,
              ('x_2_2', 'x_1_1'): -2.9257843009277966,
              ('x_2_2', 'x_1_2'): -2.925763747392389,
              ('x_2_2', 'x_2_2*x_3_1'): -400.0

In [19]:
qubo_model

{'qubo': defaultdict(float,
             {('x_3_1', 'x_3_2'): 600,
              ('x_3_1', 'x_2_1'): -4.206598246893075,
              ('x_3_1', 'x_2_2'): -4.048578424428753,
              ('x_3_1', 'x_2_1*x_1_1'): -5.435776722397996,
              ('x_3_1', 'x_2_1*x_1_2'): -5.435767766755422,
              ('x_3_1', 'x_2_2*x_1_1'): -3.63573850932578,
              ('x_3_1', 'x_2_2*x_1_2'): -3.6356660854744955,
              ('x_3_2', 'x_2_1'): -4.2291767363180615,
              ('x_3_2', 'x_2_2'): -4.07974161165229,
              ('x_3_2', 'x_2_1*x_1_1'): -5.422469200549025,
              ('x_3_2', 'x_2_1*x_1_2'): -5.422444941277003,
              ('x_3_2', 'x_2_2*x_1_1'): -3.723325784095972,
              ('x_3_2', 'x_2_2*x_1_2'): -3.7232464922542183,
              ('x_2_1', 'x_2_2'): 600,
              ('x_2_1', 'x_1_1'): 200.0,
              ('x_2_1', 'x_2_1*x_1_1'): -400.0,
              ('x_2_1', 'x_1_2'): 200.0,
              ('x_2_1', 'x_2_1*x_1_2'): -400.0,
              ('x_2

In [20]:
-2.508876295054997-2.9577370420644225-3.6357954749060384+595.9549228072192-4.2291767363180615-5.422469200549025+200.0-400.0-400.0

-22.79913194167341

In [10]:
method = 'dwave-qa'

optimizer_param = {}
optimizer_param['shots'] = 1000
optimizer_param['bucket'] = s3_bucket # the name of the bucket
optimizer_param['prefix'] = prefix # the name of the folder in the bucket
optimizer_param['device'] = "arn:aws:braket:::device/qpu/d-wave/Advantage_system4"
optimizer_param["embed_method"] = "default"

qa_optimizer = Annealer(qubo_model, method, **optimizer_param)

INFO:root:use quantum annealer arn:aws:braket:::device/qpu/d-wave/Advantage_system4 


In [11]:
# not create annealing task, only embedding logic
qa_optimizer.embed()
# create annealing task
qa_optimize_result = qa_optimizer.fit()

INFO:root:fit() ...
INFO:root:finish save /tmp/qa_result.pickle
INFO:root:_upload_result_json, bucket=amazon-braket-1a222675c751, key=annealer-experiment/d94bce7f-5670-497e-b42d-429469fcd609/qa_result.pickle
INFO:root:dwave-qa save to s3 - d94bce7f-5670-497e-b42d-429469fcd609: None


In [12]:
qa_task_id = qa_optimizer.get_task_id()
print(f"task id is {qa_task_id}")

task id is d94bce7f-5670-497e-b42d-429469fcd609


# develop post-process

In [134]:
# import networkx as nx
# from utility.MoleculeParser import MoleculeData
# from utility.QMUQUBO import QMUQUBO
# from utility.AnnealerOptimizer import Annealer
# from utility.ResultProcess import ResultParser
# from utility.MolGeoCalc import update_pts_distance
# import time
# import numpy as np

# timestamp = time.strftime("%Y%m%d-%H")
# %matplotlib inline

# %load_ext autoreload
# %autoreload 2

In [79]:
# s3_bucket = f"amazon-braket-1a222675c751" # the name of the bucket
# prefix = "annealer-experiment" # the name of the folder in the bucket
# raw_path = './molecule-data/Aspirin.mol2' # the mol2 file for this experiment
# data_path = './qmu_Aspirin_data_latest.pickle'
# qa_task_id = '58700c90-abed-48b7-bd1b-b849df22ef18'

In [13]:
method = "dwave-qa"
qa_param = {}
qa_param["bucket"] = s3_bucket
qa_param["prefix"] = prefix
qa_param["task_id"] = qa_task_id
qa_param["raw_path"] = raw_path
qa_param["data_path"] = data_path

qa_process_result = ResultParser(method, **qa_param)
# print(f"{method} result is {qa_process_result.get_all_result()}")

local_time, task_time, total_time, access_time = qa_process_result.get_time()

print(f"time for {method}: \n \
    local time is {local_time},\n \
    task time is {task_time}, \n \
    qpu total time is {total_time}, \n \
    qpu access time is {access_time}")

INFO:root:_load_raw_result
INFO:root:load quantum annealer raw result
INFO:root:_read_result_obj
INFO:root:_read_result_obj: annealer-experiment/d94bce7f-5670-497e-b42d-429469fcd609/qa_result.pickle
INFO:root:MoleculeData.load()
INFO:root:init mol data for final position
INFO:root:init mol data for raw position
INFO:root:_parse_model_info
INFO:root:_init_parameters
INFO:root:parse quantum annealer result
INFO:root:_read_result_obj
INFO:root:_read_result_obj: annealer-experiment/d94bce7f-5670-497e-b42d-429469fcd609/results.json


time for dwave-qa: 
     local time is 6.449745178222656,
     task time is 1.793, 
     qpu total time is 0.115137, 
     qpu access time is 0.101889


In [14]:
qa_atom_pos_data = qa_process_result.generate_optimize_pts()
# save unfold file for visualization and parameters for experiment: 1. volume value 2. relative improvement
qa_process_result.save_mol_file(f"{timestamp}")

INFO:root:generate_optimize_pts()
INFO:root:save_mol_file 20220311-07
INFO:root:finish save ./molecule-data/Aspirin_dwave-qa_20220311-07.mol2 and ./molecule-data/Aspirin_dwave-qa_20220311-07.json


['./molecule-data/Aspirin_dwave-qa_20220311-07.mol2',
 './molecule-data/Aspirin_dwave-qa_20220311-07.json']

In [15]:
qa_process_result.parameters

{'volume': {'optimize': 22.799131941673366,
  'initial': 22.769536671449572,
  'gain': 1.0012997748110046,
  'unfolding_results': ['x_1_1', 'x_2_1', 'x_3_2'],
  'optimize_info': {'missing_var': [], 'optimize_state': True}}}

In [16]:
print(f"valid name is {qa_process_result.valid_var_name}")

pddf_sample_result = qa_process_result.raw_result["response"].aggregate(
        ).to_pandas_dataframe()

pddf_best_result = pddf_sample_result.iloc[pddf_sample_result['energy'].idxmin(
        ), :]

best_config = pddf_best_result.filter(items=qa_process_result.valid_var_name)

best_config

valid name is ['x_3_1', 'x_3_2', 'x_2_1', 'x_2_2', 'x_1_1', 'x_1_2']


x_3_1    0.0
x_3_2    1.0
x_2_1    1.0
x_2_2    0.0
x_1_1    1.0
x_1_2    0.0
Name: 0, dtype: float64

In [17]:
result = pddf_best_result.where(pddf_best_result == 1.0)
result.dropna()

x_1_1          1.0
x_2_1          1.0
x_3_2          1.0
x_3_2*x_1_1    1.0
Name: 0, dtype: float64

In [18]:
pddf_best_result

x_1_1                    1.000000
x_1_2                    0.000000
x_2_1                    1.000000
x_2_1*x_1_1              1.000000
x_2_1*x_1_2              0.000000
x_2_2                    0.000000
x_2_2*x_1_1              0.000000
x_2_2*x_1_2              0.000000
x_3_1                    0.000000
x_3_2                    1.000000
chain_break_fraction     0.000000
energy                 -22.799132
num_occurrences         53.000000
Name: 0, dtype: float64

In [18]:
pddf_best_result

x_1_1                    1.000000
x_1_2                    0.000000
x_1_2*x_3_2              0.000000
x_2_1                    1.000000
x_2_2                    0.000000
x_2_2*x_3_1              0.000000
x_3_1                    0.000000
x_3_1*x_2_1              0.000000
x_3_2                    1.000000
x_3_2*x_1_1              1.000000
chain_break_fraction     0.000000
energy                 -22.799132
num_occurrences         44.000000
Name: 0, dtype: float64

In [ ]:
-22.79913194167341


In [82]:
import py3Dmol
import time
from ipywidgets import interact,fixed,IntSlider
import ipywidgets
import os

def View3DMol(mol, size=(600, 600), style="stick", surface=False, opacity=0.5, type="mol2"):
    assert style in ('line', 'stick', 'sphere', 'carton')
    viewer = py3Dmol.view(width=size[0], height=size[1])
    viewer.addModel(open(mol,'r').read(), type)
    viewer.setStyle({style:{}})
    if surface:
        viewer.addSurface(py3Dmol.SAS, {'opacity': opacity})
    viewer.zoomTo()
    return viewer


def View3DMolFromDir(mol_dir, size=(600,600), style="stick", surface=False, opacity=0.5, type="mol2"):
    assert style in ('line', 'stick', 'sphere', 'carton')
    viewer = py3Dmol.view(width=size[0], height=size[1],linked=False,viewergrid=(2,2))
    mol2s = os.popen("ls {}/*.mol2".format(mol_dir)).read().split('\n')[0:-1]
    for index,mol2 in enumerate(mol2s):
        viewer.addModel(open(mol2,'r').read(),type,viewer=(index/2,index%2))
        viewer.setStyle({'stick':{'colorscheme':'greenCarbon'}},viewer=(index/2,index%2))
    if surface:
        viewer.addSurface(py3Dmol.SAS, {'opacity': opacity})
    viewer.zoomTo()
    view.render()
    return viewer


def StyleSelector(mol,size,style):
    return View3DMol(mol,size=(size,size),style=style).show()

def InteractView(mol,size):
    interact(StyleSelector, 
             mol=mol,
             size=size,
             style=ipywidgets.Dropdown(
                options=['line', 'stick', 'sphere'],
                value='stick',
                description='Style:'))

In [83]:
InteractView("./molecule-data/Aspirin_dwave-qa_20220311-05.mol2", size=800)

interactive(children=(Text(value='./molecule-data/Aspirin_dwave-qa_20220311-05.mol2', description='mol'), IntS…

In [60]:
InteractView("./molecule-data/Aspirin.mol2", size=800)

interactive(children=(Text(value='./molecule-data/Aspirin.mol2', description='mol'), IntSlider(value=800, desc…

In [17]:
qa_process_result.atom_pos_data_raw

{'1': {'pts': [-1.4637, 1.3943, 1.2946],
  'idx': ([0, 0, 0], [0, 0, 0]),
  'vdw-radius': 1.7},
 '2': {'pts': [-1.3545, -0.1059, 1.201],
  'idx': ([0, 0, 0], [0, 0, 0]),
  'vdw-radius': 1.7},
 '3': {'pts': [-1.4493, -0.7819, 2.198],
  'idx': ([0, 0, 0], [0, 0, 0]),
  'vdw-radius': 1.52},
 '4': {'pts': [-1.1519, -0.6914, 0.0093],
  'idx': ([0, 0, 0], [0, 0, 0]),
  'vdw-radius': 1.52},
 '5': {'pts': [-0.9578, -2.0342, 0.0079],
  'idx': ([0, 0, 0], [0, 0, 0]),
  'vdw-radius': 1.7},
 '6': {'pts': [0.327, -2.5523, 0.0594],
  'idx': ([0, 0, 0], [0, 0, 0]),
  'vdw-radius': 1.7},
 '7': {'pts': [0.5232, -3.9195, 0.0578],
  'idx': ([0, 0, 0], [0, 0, 0]),
  'vdw-radius': 1.7},
 '8': {'pts': [-0.5586, -4.7832, 0.005],
  'idx': ([0, 0, 0], [0, 0, 0]),
  'vdw-radius': 1.7},
 '9': {'pts': [-1.843, -4.2868, -0.0466],
  'idx': ([0, 0, 0], [0, 0, 0]),
  'vdw-radius': 1.7},
 '10': {'pts': [-2.0568, -2.9052, -0.0398],
  'idx': ([0, 0, 0], [0, 0, 0]),
  'vdw-radius': 1.7},
 '11': {'pts': [-3.4263, -2.3674,

In [11]:
mol_graph = qa_process_result.mol_data.bond_graph.mol_ug

non_contact_atom_map = {}

for node_main in mol_graph.nodes:
    non_contact_atom = []
    for node_candidate in mol_graph.nodes:
        if node_candidate != node_main and node_candidate not in mol_graph.neighbors(node_main):
            non_contact_atom.append(node_candidate)
    non_contact_atom_map[node_main] = non_contact_atom

In [13]:
def calc_distance_between_pts(pts1, pts2):
    pts1_middle = np.array(tuple(list(np.mean(np.array(pts1), axis=0))))
    pts2_middle = np.array(tuple(list(np.mean(np.array(pts2), axis=0))))

    return np.linalg.norm(pts1_middle-pts2_middle)

In [21]:
atom_raw = qa_process_result.atom_pos_data

for atom_index, atom_info in atom_raw.items():
    vdw_radius = atom_info['vdw-radius']
    atom_pos = atom_info['pts']
    for non_contact_atom in non_contact_atom_map[atom_index]:
        non_contact_atom_pos = atom_raw[non_contact_atom]['pts']
        distance = calc_distance_between_pts([atom_pos], [non_contact_atom_pos])
#         print(f"fail at {atom_index} to {non_contact_atom} for distance {distance}")
        if distance < vdw_radius:
            print(f"fail at {atom_index} to {non_contact_atom}")
            break

In [26]:
a = set()

{1, 2, 3, 4}

In [20]:
graph = mol_data.bond_graph.mol_ug

In [23]:
qa_process_result.theta_option

[0.0, 45.0, 90.0, 135.0, 180.0, 225.0, 270.0, 315.0]

In [14]:
qa_process_result.mol_data.bond_graph.sort_ris_data['4']


{'4+5': {'metrics': '4+5',
  'f_0_set': {'2'},
  'f_1_set': {'10', '11', '17', '18', '19', '20', '6', '7', '8', '9'},
  'avg_bc_num': 0.5146198830409356,
  'rb_count_num': 1},
 '2+4': {'metrics': '2+4',
  'f_0_set': {'1', '3'},
  'f_1_set': {'5'},
  'avg_bc_num': 0.4444444444444444,
  'rb_count_num': 1},
 '1+2': {'metrics': '1+2',
  'f_0_set': {'14', '15', '16'},
  'f_1_set': {'3', '4'},
  'avg_bc_num': 0.36549707602339176,
  'rb_count_num': 1},
 '10+11': {'metrics': '10+11',
  'f_0_set': {'17', '18', '19', '20', '4', '5', '6', '7', '8', '9'},
  'f_1_set': {'12', '13'},
  'avg_bc_num': 0.3187134502923976,
  'rb_count_num': 1},
 '4+5,2+4': {'metrics': '4+5',
  'f_0_set': {'1', '3'},
  'f_1_set': {'10', '11', '17', '18', '19', '20', '6', '7', '8', '9'},
  'avg_bc_num': 0.47953216374269003,
  'rb_count_num': 2},
 '4+5,10+11': {'metrics': '4+5',
  'f_0_set': {'2'},
  'f_1_set': {'12', '13'},
  'avg_bc_num': 0.41666666666666663,
  'rb_count_num': 2},
 '2+4,1+2': {'metrics': '2+4',
  'f_0_se

In [15]:
qa_process_result.mol_data.bond_graph.rb_data

{'1+2': {'f_0_set': {'14', '15', '16'},
  'f_1_set': {'10',
   '11',
   '12',
   '13',
   '17',
   '18',
   '19',
   '20',
   '3',
   '4',
   '5',
   '6',
   '7',
   '8',
   '9'},
  'bc_num': 0.36549707602339176},
 '2+4': {'f_0_set': {'1', '14', '15', '16', '3'},
  'f_1_set': {'10',
   '11',
   '12',
   '13',
   '17',
   '18',
   '19',
   '20',
   '5',
   '6',
   '7',
   '8',
   '9'},
  'bc_num': 0.4444444444444444},
 '4+5': {'f_0_set': {'1', '14', '15', '16', '2', '3'},
  'f_1_set': {'10',
   '11',
   '12',
   '13',
   '17',
   '18',
   '19',
   '20',
   '6',
   '7',
   '8',
   '9'},
  'bc_num': 0.5146198830409356},
 '10+11': {'f_0_set': {'1',
   '14',
   '15',
   '16',
   '17',
   '18',
   '19',
   '2',
   '20',
   '3',
   '4',
   '5',
   '6',
   '7',
   '8',
   '9'},
  'f_1_set': {'12', '13'},
  'bc_num': 0.3187134502923976}}

In [8]:
from utility.MoleculeParser import MoleculeData
from utility.QMUQUBO import QMUQUBO
from utility.AnnealerOptimizer import Annealer
from utility.ResultProcess import ResultParser
import time

timestamp = time.strftime("%Y%m%d-%H")

2022-03-10 01:28:21,937 dwave.cloud INFO MainThread Log level for 'dwave.cloud' namespace set to 0


# Step 1: Prepare Data

In this part, we load the raw molecule data for experiment.
The [117 ligand](http://www.rcsb.org/ligand/117) was 
put in the repository. We assign the relative 
path to **raw_path**.
The **s3_bucket** and **prefix** are used to store the 
optimization results. We can use the one created with the 
cloudformation for convenience.

In [2]:
# initial parameters for experiment data
s3_bucket = f"xxxx" # the name of the bucket
prefix = "xxxx" # the name of the folder in the bucket

raw_path = './molecule-data/117_ideal.mol2' # the mol2 file for this experiment

In [3]:
mol_data = MoleculeData(raw_path, 'qmu')

data_path = mol_data.save("latest")

num_rotation_bond = mol_data.bond_graph.rb_num
print(f"You have loaded the raw molecule data and saved as {data_path}. \n\
This molecule has {num_rotation_bond} rotable bond")

INFO:root:parse mol2 file!
INFO:root:finish save qmu_117_ideal_data_latest.pickle


You have loaded the raw molecule data and saved as ./qmu_117_ideal_data_latest.pickle. 
This molecule has 23 rotable bond


After running this block, the processed data 
will be saved as **qmu_117_ideal_data_latest.pickle**
and **data_path** will be updated. We can see that this 
molecule has 23 rotatable bonds.

# Step 2: Build Model

In this part, we build the Quadratic Unconstrained 
Binary Optimization (QUBO) model for molecular unfolding.

First, we set the following parameters and 
initialize the QMUQUBO object. 

<center>

| Parameter | Description | Value |
|--- |--- |--- |
|A | penalty scalar |300|
|hubo_qubo_val | energy penalty of make_quadratic() |200|
|M | number of torsions for molecular unfolding| [1, max number of rotatable bonds] |
|D| angle precision of rotation| 8|
|method| the method of building model| 'pre-calc': calculate the score in advance|

 </center>

We use the 'pre-calc' method 
to build the model. This molecule has 23 rotatable bonds and 
we only test 2 of them, so we set the **M** to 2. And we want 
the angle to become $45^o$, so we set the **D** to 8 
(i.e., $8=360^o/45^o$). The **A** and **hubo_qubo_val** are 
test from experiments. 

In [4]:
# initial the QMUQUBO object
init_param = {}
method = ['pre-calc']

for mt in method:
    if mt == 'pre-calc':
        init_param[mt] = {}
        init_param[mt]['param'] = ['M', 'D', 'A', 'hubo_qubo_val']
    
qmu_qubo = QMUQUBO(mol_data, method, **init_param)

INFO:root:initial pre-calculate for constructing molecule QUBO


In [5]:
# set the parameters for model
model_param = {}
# parameters
num_rotation_bond = mol_data.bond_graph.rb_num

method = 'pre-calc'
model_param[method] = {}
# model_param[method]['M'] = range(1, num_rotation_bond+1)
model_param[method]['M'] = [2]
model_param[method]['D'] = [8]
model_param[method]['A'] = [300]
model_param[method]['hubo_qubo_val'] = [200]

qmu_qubo.build_model(**model_param)

INFO:root:Construct model for M:2,D:8,A:300,hubo_qubo_val:200 0.0010666688283284505 min


0

We can use the following method to check the properties of 
model. This way, we can build many models conveniently. 
After that, we save the model and update the value of 
**model_path**.

In [6]:
# describe the model parameters
model_info = qmu_qubo.describe_model()

INFO:root:method: pre-calc
INFO:root:The model_name should be {M}_{D}_{A}_{hubo_qubo_val}
INFO:root:param: M, value {2}
INFO:root:param: D, value {8}
INFO:root:param: A, value {300}
INFO:root:param: hubo_qubo_val, value {200}


In [7]:
# save the model
model_path = qmu_qubo.save("latest")

print(f"You have built the QUBO model and saved it as {model_path}")

INFO:root:finish save qmu_117_ideal_model_latest.pickle


You have built the QUBO model and saved it as ./qmu_117_ideal_model_latest.pickle


# Step 3: Optimize Configuration

In this part, we use SA and QA to find the optimized configuration of molecular unfolding.
At first, we load the model file using **QMUQUBO** object

In [8]:
qmu_qubo_optimize = QMUQUBO.load(model_path)

In [9]:
model_info = qmu_qubo_optimize.describe_model()

INFO:root:method: pre-calc
INFO:root:The model_name should be {M}_{D}_{A}_{hubo_qubo_val}
INFO:root:param: M, value {2}
INFO:root:param: D, value {8}
INFO:root:param: A, value {300}
INFO:root:param: hubo_qubo_val, value {200}


We can see the parameters of this model, with M equaling 2, D equaling 8, 
A equaling 300 and hubo_qubo_val equaling 200. 
Actually, we can contain multiple models in this file just 
by giving multiple values for one parameter when creating models.

Actually, we can contain multiple models in this file just 
by giving multiple values for one parameter when creating models.
Then, we need use **model_name** to get the model for experiments.

In [10]:
# get the model you want to optimize
M = 2
D = 8
A = 300
hubo_qubo_val = 200
model_name = "{}_{}_{}_{}".format(M, D, A, hubo_qubo_val)
method = "pre-calc"

qubo_model = qmu_qubo_optimize.get_model(method, model_name)

We can see that we want to carry out experiment with the QUBO model with M equaling 2.
 After that, we set the parameters for optimization.

| Parameter | Description | Value |
|--- |--- |--- |
|method | annealing method for QUBO problem |'dwave-sa': use the simulated annealer in ocean toolkit<br> 'dwave-qa': use the quantum annealer|
|shots| number of reads, refer to [dwave-sa](https://docs.ocean.dwavesys.com/projects/neal/en/latest/reference/generated/neal.sampler.SimulatedAnnealingSampler.sample.html#neal.sampler.SimulatedAnnealingSampler.sample) and [dwave-qa](https://amazon-braket-ocean-plugin-python.readthedocs.io/en/latest/_apidoc/braket.ocean_plugin.braket_sampler.html) for details |1 to 10,000|
|bucket | the s3 bucket to store your results | - |
|prefix | the name of the folder in your s3 bucket | - |
|device | the arn name to run your quantum annealing| 'arn:aws:braket:::device/qpu/d-wave/Advantage_system4' <br> 'arn:aws:braket:::device/qpu/d-wave/DW_2000Q_6'|

Then, we can run the SA for this problem:

In [11]:
method = 'dwave-sa'

optimizer_param = {}
optimizer_param['shots'] = 1000

sa_optimizer = Annealer(qubo_model, method, **optimizer_param)

INFO:root:use simulated annealer from dimod


In [12]:
sa_optimize_result = sa_optimizer.fit()

INFO:root:fit() ...
INFO:root:dwave-sa save to local
INFO:root:finish save sa_result.pickle


We can tell that we set the number of shots for SA to 1000. 
The result is saved as the local file **./sa_result.pickle.**
Alternatively, we can use QA to solve this problem:

In [13]:
method = 'dwave-qa'

optimizer_param = {}
optimizer_param['shots'] = 1000
optimizer_param['bucket'] = s3_bucket # the name of the bucket
optimizer_param['prefix'] = prefix # the name of the folder in the bucket
optimizer_param['device'] = "arn:aws:braket:::device/qpu/d-wave/Advantage_system4"
optimizer_param["embed_method"] = "default"

qa_optimizer = Annealer(qubo_model, method, **optimizer_param)

INFO:root:use quantum annealer arn:aws:braket:::device/qpu/d-wave/Advantage_system4 


In this QA, we set the number of shots to 1000 and 
choose the 
[Advantage_System4.1](https://docs.dwavesys.com/docs/latest/doc_physical_properties.html)
as the QPU. In addition, the results are saved to your bucket automatically and you 
can get the task id for future process. 

In [14]:
# not create annealing task, only embedding logic
qa_optimizer.embed()
# create annealing task
qa_optimize_result = qa_optimizer.fit()

INFO:root:fit() ...
INFO:root:finish save /tmp/qa_result.pickle
INFO:root:_upload_result_json, bucket=amazon-braket-1a222675c751, key=annealer-experiment/89a59fa9-c66e-482d-85e0-6dfce8c54454/qa_result.pickle
INFO:root:dwave-qa save to s3 - 89a59fa9-c66e-482d-85e0-6dfce8c54454: None


In [15]:
qa_task_id = qa_optimizer.get_task_id()
print(f"task id is {qa_task_id}")

task id is 89a59fa9-c66e-482d-85e0-6dfce8c54454


Finally, we can compare the execution time between SA and QA :

In [16]:
print(f"dwave-sa run time {sa_optimize_result['time']}")
print(f"dwave-qa run time {qa_optimize_result['time']}")

dwave-sa run time 175.73437476158142
dwave-qa run time 7.57835578918457


We can tell from the image that SA needs 174.2 seconds 
and QA needs 7.7 seconds to find 
solution.

We sometimes get the best result that occurs only once.

![OneTimeQA](../../../docs/en/images/one-time-qa.png)

This does not always indicate an error. It is actually the characteristic of the problem or how the problem 
is formulated. Because we have different linear and quadratic terms that vary by many orders of magnitude. If we 
set change value of A to some smaller number, like 10 or 100, more occurrences of the best answer will be observed. 
However, these answers usually break the constraints. For more information about this phenomenon, please refer to this 
[Link](https://support.dwavesys.com/hc/en-us/community/posts/1500000698522-Number-of-occurrences-?input_string=number%20occurance).

# Step 4: PostProcess Result

In this part, we post process the optimizing results for evaluation and visualization.
At first, we prepare the following parameters:

| Parameter | Description | Value |
|--- |--- |--- |
|method | annealing method for QUBO problem |'dwave-sa': use the simulated annealer in ocean toolkit<br> 'dwave-qa': use the quantum annealer|
|raw_path| the path for the original molecule file| './molecule-data/117_ideal.mol2' in this example |
|data_path| the path for the processed molecule file| './qmu_117_ideal_data_latest.mol2' in this example |
|bucket | the s3 bucket to store your results | - |
|prefix | the name of the folder in your s3 bucket | - |
|task_id | the id for your quantum annealing task| '2b5a3b05-1a0e-443a-852c-4ec422a10e59' in this example |

Then we can run the post-process using **ResultParser** object for SA:

In [18]:
method = "dwave-sa"
sa_param = {}
sa_param["raw_path"] = raw_path
sa_param["data_path"] = data_path

sa_process_result = ResultParser(method, **sa_param)
# print(f"{method} result is {sa_process_result.get_all_result()}")

local_time, _ , _, _= sa_process_result.get_time()

print(f"time for {method}: \n \
    local time is {local_time}")

INFO:root:_load_raw_result
INFO:root:load simulated annealer raw result
INFO:root:MoleculeData.load()
INFO:root:init mol data for final position
INFO:root:init mol data for raw position
INFO:root:_parse_model_info
INFO:root:_init_parameters
INFO:root:parse simulated annealer result
INFO:root:sa only has local_time!


time for dwave-sa: 
     local time is 175.73437476158142


In [19]:
sa_atom_pos_data = sa_process_result.generate_optimize_pts()
# save unfold file for visualization and parameters for experiment: 1. volume value 2. relative improvement
sa_process_result.save_mol_file(f"{timestamp}")


INFO:root:generate_optimize_pts()
INFO:numexpr.utils:NumExpr defaulting to 2 threads.
INFO:root:save_mol_file 20220216-06
INFO:root:finish save ./molecule-data/117_ideal_dwave-sa_20220216-06.mol2 and ./molecule-data/117_ideal_dwave-sa_20220216-06.json


['./molecule-data/117_ideal_dwave-sa_20220216-06.mol2',
 './molecule-data/117_ideal_dwave-sa_20220216-06.json']

In [20]:
sa_process_result.parameters

{'volume': {'optimize': 18.9423642473021,
  'initial': 18.549876807564765,
  'gain': 1.0211584930621898,
  'unfolding_results': ['x_15_7', 'x_14_8']}}

In the first block, we can see the **local time**
for SA is around 174 seconds. 
With the **generate_optimize_pts()** method, the final 3D 
points after unfolding will be generated and saved as json file and mol2 files. The last 
block shows the optimizing results which are also stored in json files. 
It shows that the optimized result gains 
1.0212x increase in volume. The value for **unfolding_results** indicates 
that the rotatable bond 15 should rotate $270^o$ ($360/8*(7-1)$) and 
the rotatable bond 14 should rotate $315^o$ ($360/8*(8-1)$).
At the same time, you can run the post-process for QA:

In [21]:
method = "dwave-qa"
qa_param = {}
qa_param["bucket"] = s3_bucket
qa_param["prefix"] = prefix
qa_param["task_id"] = qa_task_id
qa_param["raw_path"] = raw_path
qa_param["data_path"] = data_path

qa_process_result = ResultParser(method, **qa_param)
# print(f"{method} result is {qa_process_result.get_all_result()}")

local_time, task_time, total_time, access_time = qa_process_result.get_time()

print(f"time for {method}: \n \
    local time is {local_time},\n \
    task time is {task_time}, \n \
    qpu total time is {total_time}, \n \
    qpu access time is {access_time}")

INFO:root:_load_raw_result
INFO:root:load quantum annealer raw result
INFO:root:_read_result_obj
INFO:root:_read_result_obj: annealer-experiment/89a59fa9-c66e-482d-85e0-6dfce8c54454/qa_result.pickle
INFO:root:MoleculeData.load()
INFO:root:init mol data for final position
INFO:root:init mol data for raw position
INFO:root:_parse_model_info
INFO:root:_init_parameters
INFO:root:parse quantum annealer result
INFO:root:_read_result_obj
INFO:root:_read_result_obj: annealer-experiment/89a59fa9-c66e-482d-85e0-6dfce8c54454/results.json


time for dwave-qa: 
     local time is 7.57835578918457,
     task time is 1.89, 
     qpu total time is 0.142533, 
     qpu access time is 0.124129


we can see that there many types of time metrics for running QA.
This task has the **local time** of 7.7 s, which means the time between calling the api and 
getting the annealing result. The **task time** time is the metric from the json file in 
bucket. We can also see the **qpu total time** and **qpu access time** representing the 
actual time running in the QPU. Please refer to [Operation and Timing](https://docs.dwavesys.com/docs/latest/c_qpu_timing.html)
for details.

In [22]:
qa_atom_pos_data = qa_process_result.generate_optimize_pts()
# save unfold file for visualization and parameters for experiment: 1. volume value 2. relative improvement
qa_process_result.save_mol_file(f"{timestamp}")

INFO:root:generate_optimize_pts()
INFO:root:save_mol_file 20220216-06
INFO:root:finish save ./molecule-data/117_ideal_dwave-qa_20220216-06.mol2 and ./molecule-data/117_ideal_dwave-qa_20220216-06.json


['./molecule-data/117_ideal_dwave-qa_20220216-06.mol2',
 './molecule-data/117_ideal_dwave-qa_20220216-06.json']

In [23]:
qa_process_result.parameters

{'volume': {'optimize': 18.9423642473021,
  'initial': 18.549876807564765,
  'gain': 1.0211584930621898,
  'unfolding_results': ['x_15_7', 'x_14_8']}}

In same way, the optimized results are translated the 3D points and saved 
as local json and mol2 files. The result indicates that QA gains 
1.021x increase in 
volume.

Finally, We can open folders for the optimized results:

![optimize-results](../../../docs/en/images/optimize-results.png)

 <center>Optimize Results</center>

We can see the json result and mol2 file of SA and QA are 
stored in this place. If we carry out more 
experiments, more results with time stamp are 
stored incrementally. 
For visualization, 
we can upload the 
result **117_ideal_dwave-qa_20220216-05.mol2** 
into 
[online viewer tool](https://www.rcsb.org/3d-view) 
to see the result:

![visual](../../../docs/en/images/visualization.png)

 <center>Visualization</center>